src/main/java/com/scaleunlimited/flinkcrawler/urls/SimpleUrlLengthener.java

```diff
@@ -1,21 +1,118 @@
+
 @SuppressWarnings("serial")
 public class SimpleUrlLengthener extends BaseUrlLengthener {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SimpleUrlLengthener.class);
 
-	@Override
+    private static final Pattern HOSTNAME_PATTERN = Pattern.compile("^http://([^/:?]{3,})");
+
+    private BaseHttpFetcher _fetcher;
+    private Set<String> _urlShorteners;
+
+    public SimpleUrlLengthener(BaseHttpFetcherBuilder fetcherBuilder) {
+        super();
+        try {
+            _fetcher = fetcherBuilder.build();
+            _urlShorteners = loadUrlShorteners();
+        } catch (Exception e) {
+            throw new RuntimeException("Unable to build URL lengthener", e);
+        }
+    }
+	
+    @Override
 	public RawUrl lengthen(RawUrl url) {
 		// If the domain is a link shortener, lengthen it.
 		// FUTURE  Try to fetch the URL. This call needs to be thread-safe
 		// See https://github.com/ScaleUnlimited/flink-crawler/issues/50
 		
-		return url;
+        String urlString = url.getUrl();
+        
+        Matcher m = HOSTNAME_PATTERN.matcher(urlString);
+        if (!m.find()) {
+            return url;
+        }
+
+        String hostname = m.group(1);
+        if (!_urlShorteners.contains(hostname)) {
+            // FUTURE - see if this looks like a shortened URL
+            return url;
+        }
+        
+        String redirectedUrl = urlString;
+
+        try {
+            FetchedResult fr = _fetcher.get(urlString);
+            int statusCode = fr.getStatusCode();
+            if (statusCode == HttpStatus.SC_OK) {
+                redirectedUrl = fr.getFetchedUrl();
+                LOGGER.debug(String.format("Normal redirection of %s to %s", urlString, redirectedUrl));
+            } else {
+                LOGGER.trace("Status code " + statusCode + " processing redirect for " + urlString);
+            }
+        } catch (RedirectFetchException e) {
+            // We'll get this exception if the URL that's redirected by
+            // a link shortening site is to a URL that gets redirected again.
+            // In this case, we've captured the final URL in the exception,
+            // so use that for downstream fetching.
+            redirectedUrl = e.getRedirectedUrl();
+            LOGGER.trace(String.format("Redirecting %s to %s", urlString, redirectedUrl));
+        } catch (BaseFetchException e) {
+            // We might have hit a site that doesn't process HEAD requests properly,
+            // so just emit the same URL for downstream fetching.
+            LOGGER.debug("Exception processing redirect for " + urlString + ": " + e.getMessage(), e);
+        }
+        
+		try {
+            return new RawUrl(redirectedUrl, url.getScore());
+        } catch (MalformedURLException e) {
+            LOGGER.debug(urlString + " redirected to malformed URL: " + redirectedUrl);
+            return url;
+        }
 	}
 
 	@Override
 	public int getTimeoutInSeconds() {
 		return 10;
```

Comment:
This should be based on the fetcher's timeout.

Reply:
Done